org.jacoco.core.test/src-java7/org/jacoco/core/test/filter/TryWithResourcesTest.java

```diff
@@ -0,0 +1,175 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.filter;
+
+import java.io.IOException;
+
+import org.jacoco.core.analysis.ICounter;
+import org.jacoco.core.internal.Java9Support;
+import org.jacoco.core.test.TargetLoader;
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.filter.targets.TryWithResources;
+import org.junit.Test;
+
+/**
+ * Test of filtering of a bytecode that is generated for a try-with-resources
+ * statement.
+ */
+public class TryWithResourcesTest extends ValidationTestBase {
+
+	public TryWithResourcesTest() {
+		super("src-java7", TryWithResources.class);
+	}
+
+	/**
+	 * {@link TryWithResources#test()}
+	 */
+	@Test
+	public void test() {
+		assertLine("test.before", ICounter.FULLY_COVERED);
+		// without filter next line covered partly:
+		assertLine("test.try", ICounter.FULLY_COVERED);
+		assertLine("test.open1", ICounter.FULLY_COVERED);
+		assertLine("test.open2", ICounter.FULLY_COVERED);
+		assertLine("test.open3", ICounter.FULLY_COVERED);
+		assertLine("test.body", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		assertLine("test.close", ICounter.EMPTY);
+		assertLine("test.catch", ICounter.NOT_COVERED);
+		assertLine("test.finally", ICounter.PARTLY_COVERED);
+	}
+
+	/**
+	 * {@link TryWithResources#test2()}
+	 */
+	@Test
+	public void test2() {
+		assertLine("test2.before", ICounter.FULLY_COVERED);
+		// without filter next line covered partly:
+		assertLine("test2.try", ICounter.FULLY_COVERED);
+		assertLine("test2.open1", ICounter.FULLY_COVERED);
+		assertLine("test2.open2", ICounter.FULLY_COVERED);
+		assertLine("test2.open3", ICounter.FULLY_COVERED);
+		assertLine("test2.body", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		assertLine("test2.close", ICounter.EMPTY);
+		assertLine("test2.catch", ICounter.NOT_COVERED);
+		assertLine("test2.finally", ICounter.PARTLY_COVERED);
+		assertLine("test2.after", ICounter.FULLY_COVERED);
+	}
+
+	/**
+	 * {@link TryWithResources#returnInBody()}
+	 */
+	@Test
+	public void returnInBody() {
+		// without filter next line covered partly:
+		assertLine("returnInBody.try", ICounter.FULLY_COVERED);
+		assertLine("returnInBody.open", ICounter.FULLY_COVERED);
+		// without filter next line has branches:
+		// TODO with javac 1.8.0_31 part of return instructions belong to close
+		// assertLine("returnInBody.close", ICounter.EMPTY);
```

Comment:
In case of

```java
private static Object returnInBody() throws IOException {
    try ( // $line-returnInBody.try$
        Closeable r = new Resource() // $line-returnInBody.open$
    ) {
        return read(r); // $line-returnInBody.return$
    } // $line-returnInBody.close$
}
```

we correctly find instructions, but javac 7 as well as javac 8 up to 8u77 (tested public updates) generates:

```
LINENUMBER $line-returnInBody.return$
...
ASTORE
LINENUMBER $line-returnInBody.close$
... // close resource
ALOAD
ARETURN
LINENUMBER $line-returnInBody.try$
... // save exception
LINENUMBER $line-returnInBody.close$
... // close resource
```

while javac starting from 8u92 (notice additional `LINENUMBER` before `ALOAD`) as well as latest 9 EA:

```
LINENUMBER $line-returnInBody.return$
...
ASTORE
LINENUMBER $line-returnInBody.close$
... // close resource
LINENUMBER $line-returnInBody.return$
ALOAD
ARETURN
LINENUMBER $line-returnInBody.try$
... // save exception
LINENUMBER $line-returnInBody.close$
... // close resource
```

Looks like this relates to [JDK-8134759](https://bugs.openjdk.java.net/browse/JDK-8134759) (log of bisect - [bisect.txt](https://github.com/jacoco/jacoco/files/892452/bisect.txt))

There is no such problem when `finally` is present, but this case is covered by another existing test.

So I'm wondering what can we do with this test?

Travis provides JDK 8u31 by default and we also test JDK 8 EA. Can imagine that for the bytecode version 7 generated by javac we will be testing old behavior, while will be testing and using updated stable JDK 8 in Travis as well as locally.

But maybe @marchof you have some other thoughts/ideas?